packages/core/src/protocol/Invoice.ts

```diff
@@ -0,0 +1,342 @@
+ * Sub object describing receiving currency in more detail
+ *
+ * @param name
+ * @param code
+ * @param symbol
+ * @param decimals
+ */
+export type InvoiceCurrency = z.infer<typeof InvoiceCurrencySchema>;
+
+const InvoiceSchema = z.object({
+  receiverUma: z.string(),
+  invoiceUUID: z.string().uuid(),
+  amount: z.number(),
+  receivingCurrency: InvoiceCurrencySchema,
+  expiration: z.number(),
+  isSubjectToTravelRule: z.boolean(),
+  requiredPayerData: optionalIgnoringNull(CounterPartyDataOptionsSchema),
+  umaVersion: z.string(),
+  commentCharsAllowed: optionalIgnoringNull(z.number()),
+  senderUma: optionalIgnoringNull(z.string()),
+  invoiceLimit: optionalIgnoringNull(z.number()),
+  kycStatus: optionalIgnoringNull(z.nativeEnum(KycStatus)),
+  callback: z.string(),
+  signature: optionalIgnoringNull(z.instanceof(Uint8Array)),
+});
+
+/**
+ * Invoice
+ * represents a UMA invoice
+ *
+ * @param receiverUma
+ * @param invoiceUUID - Invoice UUID Served as both the identifier of the UMA invoice, and the validation of proof of payment
+ * @param amount - The amount of invoice to be paid in the smallest unit of the ReceivingCurrency.
+ * @param receivingCurrency - The currency of the invoice
+ * @param expiration - The unix timestamp of when the UMA invoice expires
+ * @param isSubjectToTravelRule -  Indicates whether the VASP is a financial institution that requires travel rule information.
+ * @param requiredPayerData - the data about the payer that the sending VASP must provide in order to send a payment
+ * @param umaVersion - UmaVersion is a list of UMA versions that the VASP supports for this transaction. It should be
+ * containing the lowest minor version of each major version it supported, separated by commas
+ * @param commentCharsAllowed - is the number of characters that the sender can include in the comment field of the pay request.
+ * @param senderUma - The sender's UMA address. If this field presents, the UMA invoice should directly go to the sending VASP
+ * instead of showing in other formats
+ * @param invoiceLimit - The maximum number of the invoice can be paid
+ * @param kycStatus - YC status of the receiver, default is verified
+ * @param callback - The callback url that the sender should send the PayRequest to
+ * @param signature - The signature of the UMA invoice
+ */
+export type Invoice = z.infer<typeof InvoiceSchema>;
+
+type TLVSerial<T> = {
+  tag: number;
+  serialize: (value: T) => Uint8Array;
+  deserialize: (bytes: Uint8Array) => T;
+};
+
+/**
+ * Serializer object converts InvoiceCurrency to Uint8Array,
+ * or creates InvoiceCurrency based on properly validated Uint8Array
+ */
+const TLVInvoiceCurrencySerializer = {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  serialMap: new Map<string, TLVSerial<any>>(),
+
+  reverseLookupSerialMap: new Map<number, string>(),
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  registerTLV(field: string, helper: TLVSerial<any>) {
```

Comment:
Opened a PR [here](https://github.com/uma-universal-money-address/uma-js-sdk/pull/117) that gets rid of most of these `any` in favor of `unknown`. `any` should generally not be used because it turns of typechecking for that entity which can introduce lots of type issues. Instead if I don't know the type I use `unknown` which forces you to assert the type (often with a type guard) where it's used.

This particular any (registerTLV) is pretty tricky to resolve. I tried making it all possible function types but the problem is when you call the function the parameter is forced to be `never` because it's too broad. I think the best way of handling this, if it works for you, would be a catch all serializer / deserializer that looks something like:
```
function serialize(arg: unknown) {
  if (typeof arg === "string") {
    serializeString(arg);
  } else if (typeof arg === "boolean") {
    serializeBoolean(arg);
  } else if (typeof arg === "number") {
    serializeNumber(arg);
  } else if (isCounterPartyDataOptions(arg)) {
    counterPartyDataOptionsToBytes(arg);
  } .... etc
}
```
I think isCounterPartyDataOptions could just use `CounterPartyDataOptionsSchema.safeParse`, something like:
```
function isCounterPartyDataOptions(value: unknown): value is CounterPartyDataOptions {
  const result = CounterPartyDataOptionsSchema.safeParse(value);
  return result.success;
}
```